### PR TITLE
[#3] Attemp to impl Cached binary search

### LINKED ISSUE #3: Use cache-friendly binary search
You can try to use different implementation of binary search algorithm (though still very simple). You can use cache-friendly binary search:

* http://bannalia.blogspot.ru/2015/06/cache-friendly-binary-search.html

This can improve performance of lookup. I guess you can create another module and implement it there. In order to implement this algorithm you need to perform two steps:

1. Implement `Vector` initialization from sorted list.
2. Implement cache-friendly binary search.

2 is simple, because algo itself is simple and there's C code in blog post. Though, 1 is a little bit more difficult...

1. First, you need to sort list.
2. Then you need to create possibly perfectly-balanced binary tree from this sorted list. Introduce your own 
```haskell
data Tree a = Leaf | Node a (Tree a) (Tree a)
```
3. Convert `Tree` to list using level-order.
4. Finally, just create `Vector` from this list.

So, probably initialization of `Vector` becomes slower. But AFAIU, this is not important. 

The only drawback: if first halves of `Fingerpint` are equal, you can't just increment index by 1 `i + 1` to search for the next key. But, assuming this is not important we can forget about this :) Or actually compare by second part of `Fingerptint` in case first part is equal. You can just print number of unique first halves of `Fingeprint`s in benchmark first (using just `print`) to see how many different elements there.

### PR DESCRIPTION
**Pull request by vrom911**
Resolves #3 

As seen from benchmarks this implementation is a bit faster

```
benchmarking map/lookup
time                 2.314 μs   (2.307 μs .. 2.324 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 2.309 μs   (2.305 μs .. 2.316 μs)
std dev              15.68 ns   (10.99 ns .. 22.24 ns)

benchmarking vector optimal cache/lookup
time                 340.2 ns   (339.5 ns .. 340.9 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 339.7 ns   (339.1 ns .. 340.3 ns)
std dev              1.857 ns   (1.588 ns .. 2.267 ns)

benchmarking vector optimal/lookup
time                 447.2 ns   (446.5 ns .. 447.9 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 447.0 ns   (446.5 ns .. 448.2 ns)
std dev              2.311 ns   (1.607 ns .. 3.839 ns)

```

Also, I tried to use `W64#` in optimal vector implementation but it didn't affect the performance.

### CODE REVIEW

**Comment by int-index** [comment]
If I read the report right, it's a 24% perf improvement. Nice!

**Comment by ChShersh** [comment]
@int-index Implementation of cached binary search really changes internals of structure and this makes `insert` and data structure traversals slower so this data structure is not really dynamic. Is it okay? What use cases for such `typerep-map` do you have in mind?

File: src/Data/TypeRep/CacheMap.hs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+cachedBinarySearch :: Fingerprint -> Unboxed.Vector Word64 -> Unboxed.Vector Word64 -> Maybe Int
+cachedBinarySearch (Fingerprint (W64# a) (W64# b)) fpAs fpBs = inline (go 0#)
+  where
+    go :: Int# -> Maybe Int
+    go i = case i <# len of
+        0# -> Nothing
+        _  -> let !(W64# valA) = Unboxed.unsafeIndex fpAs (I# i) in case a `ltWord#` valA of
+            0#  -> case a `eqWord#` valA of
+                0# -> go (2# *# i +# 2#)
+                _ -> let !(W64# valB) = Unboxed.unsafeIndex fpBs (I# i) in case b `ltWord#` valB of
```
Comment (ChShersh) [inline src/Data/TypeRep/CacheMap.hs]: I think this part can be optimized a bit. In almost all cases (as far as I understand) if first halves of of fingreprint are equal, second halves will be equal as well. So we don't need to perform extra redundant `ltWord#` in almost all cases. And thus we can check for `eqWord#` first.

File: src/Data/TypeRep/CacheMap.hs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+fromListToTree xs =
+    let len = length xs
+    in case splitAt (len `div` 2) xs of
+           ([], [])     -> Leaf
+           (ls, [])     -> Node (last ls) (fromListToTree $ init ls) Leaf
+           ([], (r:rs)) -> Node r Leaf $ fromListToTree rs
+           (ls, r:rs)   -> Node r (fromListToTree ls) (fromListToTree rs)
+
+breadthFirst :: Tree a -> [a]
+breadthFirst tree = bf [tree]
```
Comment (ChShersh) [inline src/Data/TypeRep/CacheMap.hs]: Nice implementation!

File: src/Data/TypeRep/CacheMap.hs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+data Tree a = Leaf | Node a (Tree a) (Tree a)
+    deriving (Show)
+
+fromListToTree :: [a] -> Tree a
+fromListToTree [] = Leaf
+fromListToTree xs =
+    let len = length xs
+    in case splitAt (len `div` 2) xs of
+           ([], [])     -> Leaf
```
Comment (ChShersh) [inline src/Data/TypeRep/CacheMap.hs]: Strictly speaking, this case can never happen. But we can't prove this to compiler. So it's okay to check this.

File: src/Data/TypeRep/CacheMap.hs (hunk 1 above)
Comment (vrom911) [reply src/Data/TypeRep/CacheMap.hs]: Thanks!
It's improved for the tiny amount (~ 6 ns)

**Comment by vrom911** [comment]
@int-index @ChShersh, I was trying to check if comparison on the second `Word64` of fingerprint takes a lot of time so I commented that part of the code completely (so binary search only on one vector). And the strange part is that benchmarks show slower results :thinking: Do you have any idea how that happens?

**Comment by vrom911** [comment]
I didn't notice anything that can explain this behaviour. Here is the generated code of both cases:
https://gist.github.com/vrom911/6dbd65e8c47f89d468a4bd225b1b5a8d

Anyway, I'm going to make this solution the main one as it gives some advantages.

**Comment by ChShersh** [comment]
@vrom911 This looks strange to me. And looks like GHC bug (or some weirdness in CPU). It doesn't make sense to me why code with more checks takes less time. Both generated cores looks like direct translation of your code. There's nothing suspicious.

**Comment by vrom911** [comment]
😢 When I'm putting it in the main library from internal one then the generated core grows and becomes less-performant (but still better than OptimalVector) as I see in benchmarks. 😞How is that even possible? I'm just renaming files.

**Comment by vrom911** [comment]
Oh, my. I found the solution! 
I need to move all other modules into `internal/` folder because for some reason it doesn't work when I'm using the same `hs-source-dirs` both for internal and main library and the main one depends on the internal lib. 
Btw, as far as I see from the issues at stack repo, it won't work with stack because of the internal library.

**Comment by ChShersh** [comment]
@vrom911 These are some crazy tricks to make it work...

**Comment by vrom911** [comment]
Wow, just look at these! 😮 
```
benchmarking vector optimal cache/lookup
time                 256.5 ns   (256.2 ns .. 256.9 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 256.7 ns   (256.4 ns .. 257.3 ns)
std dev              1.432 ns   (773.1 ps .. 2.217 ns)
```

**Comment by int-index** [comment]
I'd like to see the results with `{-# NOINLINE tenLookups #-}`. I wonder if inlining messes around with the benchmarking framework itself.

**Comment by vrom911** [comment]
```
benchmarking vector optimal cache/lookup
time                 273.6 ns   (273.2 ns .. 274.0 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 273.5 ns   (273.1 ns .. 274.0 ns)
std dev              1.450 ns   (942.5 ps .. 2.328 ns)
```

But I also checked the generated core and it totally different from the previous ones.

**Comment by int-index** [comment]
`273.6 ns` with `{-# NOINLINE #-}`, correct? That's an incredible result. A solid 1.5x improvement.

**Comment by vrom911** [comment]
Yes, that's correct. I didn't expect results like this 🎉

**Pull request merged**